spec/lang/well-formed.md

```diff
@@ -434,10 +438,26 @@ impl Terminator {
             Goto(block_name) => {
                 list![block_name]
             }
-            If { condition, then_block, else_block } => {
-                let ty = condition.check_wf::<T>(live_locals, prog)?;
-                ensure(matches!(ty, Type::Bool))?;
-                list![then_block, else_block]
+            Switch { value, cases, fallback } => {
+                let ty = value.check_wf::<T>(live_locals, prog)?;
+                let Type::Int(switch_ty) = ty else {
+                    // We only switch on integers.
+                    // This is in contrast to Rust MIR where switch can work on `char`s and booleans as well.
+                    // However since those are trivial casts we chose to only accept integers.
+                    throw!()
+                };
+
+                // gets the successor blocks of all cases that are valid.
+                let mut next_blocks = cases.iter().filter_map(|(case, block)| {
+                    ensure(switch_ty.can_represent(case))?;
+                    Some(block)
+                }).collect::<List<BbName>>();
+
+                // ensures that all cases were valid.
+                ensure(next_blocks.len() == cases.len())?;
```

Comment:
A `for` loop would be easier to understand than doing a `filter_map` with a length check.